### PR TITLE
Defence Range widget optimizations

### PR DESCRIPTION
**Pull request by esainane**

### CODE REVIEW

File: LuaUI/Widgets/gui_defenseRange.lua [hunk 1]
```diff
@@ -585,15 +649,16 @@ local function SetupChiliStuff()
...
-		local unitAllyTeam = Spring.GetUnitAllyTeam(unitID)
-		UnitDetected(unitID, Spring.GetUnitDefID(unitID), unitAllyTeam == myAllyTeam)
+	DoFullUnitReload()
+end
+
+function widget:Shutdown()
+	for unitID,def in pairs(defences) do
+		glDeleteList(def.drawList)
 	end
 end
```
Comment (sprunk) [inline LuaUI/Widgets/gui_defenseRange.lua:664]:
```suggestion
	end
	gl.DeleteList(defenseRangeDrawList)
end
```

File: LuaUI/Widgets/gui_defenseRange.lua [hunk 2]
```diff
@@ -448,33 +479,66 @@ function widget:Update(dt)
...
 		if unitDefID then
-			if defences[unitID].checkCompleteness then
-				-- Not allied because this check is only required for enemies.
-				-- Allied units are detected in UnitFinished.
-				UnitDetected(unitID, unitDefID, false)
+			UnitDetected(unitID, unitDefID, false)
+		end
+	end
+
+	for unitID in pairs(defNeedingLosChecks) do
```
Comment (sprunk) [inline LuaUI/Widgets/gui_defenseRange.lua:507]:
```suggestion
	for unitID in pairs(defNeedingLosChecks) do -- TODO: rarely updated but constantly iterated, consider IndexableArray
```

File: LuaUI/Widgets/gui_defenseRange.lua [hunk 3]
```diff
@@ -448,33 +479,66 @@ function widget:Update(dt)
...
+		if not spGetUnitDefID(unitID) then
+			local def = defences[unitID]
+			if select(2, spGetPositionLosState(def.x, def.y, def.z)) then
```
Comment (sprunk) [inline LuaUI/Widgets/gui_defenseRange.lua:510]: select is slightly slower than locals 
```suggestion
			local _, inLos = spGetPositionLosState(def.x, def.y, def.z)
			if inLos then
```

File: LuaUI/Widgets/gui_defenseRange.lua [hunk 4]
```diff
@@ -439,42 +468,74 @@ function widget:Update(dt)
...
+	end
+end
+
 function widget:GameFrame(n)
 	if (n % 30) ~= 0 then
 		return
 	end
 
-	for unitID, def in pairs(defences) do
+	for unitID in pairs(defNeedingBuildingChecks) do
```
Comment (GoogleFrog) [inline LuaUI/Widgets/gui_defenseRange.lua:496]: I would prefer that these were not pairs.

**Pull request merged**